proposals/type-imports/Overview.md

```diff
@@ -4,35 +4,79 @@
+In order to maintain encapsulation of such imports, even when defined in Wasm itself, the proposal further introduces the notion of *private* type definitions, which are opaque when exported.
+
+
+### Design Rationale
+
+As far as a Wasm module is concerned, _imported_ types are abstract.
+Due to Wasm's staged compilation/instantiation model, an imported type's definition is not known at compile time.
+
+However, an import may specify a subtype constraint by giving a supertype *bound* with the import (partial abstraction).
+Such an import can only be instantiated by a type that actually is a subtype of the bound.
+The type `any` can be used if no constraint is desired;
+such an import can be instantiated with both an `extern` type provided by the host or a type implemented in Wasm itself.
+
+Type _exports_ are transparent by default.
+When using a type export to instantiate the import of another module,
+its full definition is therefore available to verify any import constraints.
+
+However, it also ought to be possible to make type exports opaque,
+in order to *encapsulate* their definition (akin to an abstract data type).
+There are several requirements for opaque type definitions (called "private" in this proposal):
+
+* Encapsulation must be both static and dynamic. In particular, a cast (like the runtime type check for call_indirect, or a down cast as in the GC proposal) must not pierce through the encapsulation barrier.
+
+* Encapsulation must also be sustained when values are passed to a high-level embedding language, such as JavaScript.
+
+* Encapsulated values must be compatible with unconstrained imports, i.e, type `any`, such that a private type can be used to implement a type import.
+
+* Yet, encapsulation must not get in the way of runtime type checks. For example, it must remain possible to `call_indirect` a function with a private type among its parameters (this essentially is a higher-order cast).
+
+* In a similar vein, encapsulation must be forward-compatible with the addition of explicit casts. It ought to be possible to inject an encapsulated value into `anyref` and cast it back, in order for private types to participate in the same anyref-based type escape hatches as other references, and to avoid more complicated type system machinery. That is, it must be possible to compare with or cast _to_ a private type, but not _through_ it.
+
+Together, these requirements and goals necessitate that (1) private types are nominal, in order to distinguish them from one another, (2) values of private type share a representation with other reference types, in order to make private types suitable for imports, and (3) these values have a distinguished representation that maintains enough type information to distinguish them from values of other type.
+The latter in turn necessitates that such values are allocated -- as is typically the case for external, host-implemented references as well.
```

Comment:
I'd nuance this to explain that this boxing is only necessary for the current set of value types; with future GC/function reference types, when a module can control the type definition, it can avoid boxing by putting the privacy flag in the type definition.

Reply:
Done.